keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -589,6 +618,83 @@ associativeList.all(e, e.key1 == 'a' && e.key2 == 'b' && e.val == 100)
 associativeList.all(e, e.val == 100)
 ```
 
+### Access to Existing Object State
+
+#### Use Cases
+
+| Use Case                                                          | CEL Rule                                                                        | Examples |
+| ----------------------------------------------------------------- | ------------------------------------------------------------------------------- | -------- |
+| Immutability                                                      | `self.foo == oldSelf.foo`                                                       ||
+| Prevent modification/removal once assigned                        | `oldSelf != 'bar' || self == 'bar'` or `!has(oldSelf.field) || has(self.field)` | [featureSet selection](https://github.com/openshift/kubernetes/blob/b426a3ffde5e7cd78483adeb5cb855f9f04dec5a/openshift-kube-apiserver/admission/customresourcevalidation/features/validate_features.go#L81-L88) |
+| Append-only set                                                   | `self.all(element, element in oldSelf)`                                         ||
+| If previous value was X, new value can only be A or B, not Y or Z | `oldSelf != 'X' || self in ['A', 'B']`                                          | [Latency profile selection](https://github.com/openshift/enhancements/pull/977) |
+| Tolerating once-acceptable values ("ratcheting")                  | `self == oldSelf || self in ["currently", "acceptable", "values"]`              ||
```

Comment:
the ratcheting example is interesting... more details (both the `rule` and `updateRule` content) for a couple different scenarios might help clarify the interaction between those fields

Imagine a known-bad value `badValue` in a field `foo` inside a list of type `map` (keyed by a `name` field, for example). What `rule`/`updateRule` combination would:
1. prevent creating a new object with `foo: badValue`
2. prevent a `{"name":"a", "foo":"goodValue"}` → `{"name":"a", "foo":"badValue"}` update
3. prevent a `{"name":"a", "foo":"goodValue"}` → `{"name":"b", "foo":"badValue"}` update
4. tolerate a `{"name":"a", "foo":"badValue"}` → `{"name":"a", "foo":"badValue", "other":"value"}` update

Reply:
```yaml
x-kubernetes-validations:
- rule: isGoodValue
  ratcheting: true
- rule: someThingToBeTrueEvenInOldObjects
  ratcheting: false
```

Reply:
@liggitt Assuming a given item in the array compares against an item in the existing array based on the key, I think it the schema would look something like:

```yaml
type: object
properties:
  widgets:
    type: array
    x-kubernetes-list-type: map
    x-kubernetes-list-map-keys: ["name"]
    items:
      type: object
      properties:
        name:
          type: string
        foo:
          type: string
          x-kubernetes-validations:
          - rule: "self == 'good'"
            updateRule: "self in (oldSelf == 'bad' ? ['bad', 'good'] : ['good'])"
        bar:
          type: string
```
With decisions:
```
REJECT (not comparable: 'rule')   {"widgets":[]} => {"widgets":[{"name": "a", "foo": "bad"}]}
REJECT (comparable: 'updateRule') {"widgets":[{"name": "a", "foo": "good"}]} => {"widgets":[{"name": "a", "foo": "bad"}]}
REJECT (not comparable: 'rule')   {"widgets":[{"name": "a", "foo": "good"}]} => {"widgets":[{"name": "b", "foo": "bad"}]}
ACCEPT (comparable: 'updateRule') {"widgets":[{"name": "a", "foo": "bad"}]} => {"widgets":[{"name": "a", "foo": "bad", "bar": "hello"}]}
```